### PR TITLE
Enhance LawnMower State Management and Error Handling

### PR DESCRIPTION
**Pull request by ClockeNessMnstr**
### **User description**
Luba was getting into some weird states when trying to resume while returning. 
Not sure how I did on the exception handling. 

I explicitly checked the mode/status before executing the commands. 
I handled:
 resuming from returning.
 starting from returning.
 pausing from returning. 
 
made sure the "finally" block was run. 
I believe it was possible to miss the finally block in some cases with errors or returns with multiple try blocks. 

Also added "RETURNING" status from Lawnmower to match implementation from HA and i found it disconcerting just seeing "Mowing" as the status while luba returned.

___

### **Description**
- Introduced `RETURNING` state for the lawn mower to improve status reporting.
- Implemented cancellation of return to dock when the mower is starting or resuming from returning.
- Enhanced error handling with specific failure messages for pause and docking operations.
- Limited command execution based on the mower's current work mode to prevent invalid operations.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>lawn_mower.py</strong><dd><code>Enhance LawnMower State Management and Error Handling</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

custom_components/mammotion/lawn_mower.py
<li>Added handling for <code>RETURNING</code> mode in mower activity.<br> <li> Cancel return to dock when starting or resuming from returning mode.<br> <li> Improved exception handling with specific translation keys for errors.<br> <li> Limited commands based on mower's work mode.<br>

</details>

  </td>
  <td><a href="https://github.com/mikey0000/Mammotion-HA/pull/135/files#diff-3af9b47c563d00133985b13d37c0fb657cc947de1f4c7feae8bdeb8d849de9d7">+92/-51</a>&nbsp; </td>
</tr>
</table></td></tr></tr></tbody></table>

### CODE REVIEW

**Comment by ClockeNessMnstr** [comment]
Just saw #112 for the "returning" state lol

reverting that part.

**Comment by ClockeNessMnstr** [comment]
Is MODE_READY possible to exist with an active task? 

From watching it a little I thought MODE_READY would mean no active task.

**Comment by ClockeNessMnstr** [comment]
took off the check for work area and updated "mode" after cancelling return to charge. 
works well although the preview in the app lost the path momentarily. not sure if that was related.

File: custom_components/mammotion/lawn_mower.py [hunk 1]
```diff
@@ -85,71 +84,110 @@ async def async_start_mowing(self) -> None:
...
             try:
-                await self.coordinator.async_send_command("resume_execute_task")
-                return await self.coordinator.async_request_iot_sync()
+                if mode == WorkMode.MODE_RETURNING:
+                    trans_key = "dock_failed"
+                    await self.coordinator.async_send_command("cancel_return_to_dock")
+                    await self.coordinator.async_request_iot_sync()
+                    mode = self.rpt_dev_status.sys_status
+                if mode == WorkMode.MODE_PAUSE :
+                    trans_key = "resume_failed"
```
Comment (mikey0000) [inline custom_components/mammotion/lawn_mower.py:106]: If the mower is on the charger recharging its in WorkMode.MODE_READY but needs resumed, for example I resume mine early sometimes at 80% charge

**Comment by mikey0000** [comment]
I'll give this a review later on, thanks for the contribution.

File: custom_components/mammotion/lawn_mower.py (hunk 1 above)
Comment (ClockeNessMnstr) [reply custom_components/mammotion/lawn_mower.py:107]: Mine is definitely in "MODE_PAUSE" when recharging.
image

Comment (ClockeNessMnstr) [reply custom_components/mammotion/lawn_mower.py:107]: Conversely work area does not update to "0" when task is cancelled, if that is what "area" was checking. 

image

**Comment by ClockeNessMnstr** [comment]
#140 seemed almost related enough to include here but made a separate PR. 
Looks like that one needs translations for its name to look nice in automations. 

I could add exception keys for dock_cancel, cancel as well if that's useful but they would be translations that may not be seen. 
If you are just doing translations from google translate or something I could fill those out the same way though.

**Comment by ClockeNessMnstr** [comment]
merged in the service call to perform state checks before planning/starting. 
this PR will resume instead of starting the new task if a task is already running when service call is executed. 
However, #140 will cancel the first and run the new task which is more likely desired.

File: custom_components/mammotion/lawn_mower.py [hunk 2]
```diff
@@ -166,77 +166,104 @@ async def async_start_mowing(self, **kwargs: Any) -> None:
...
+                WorkMode.MODE_READY,
+                WorkMode.MODE_RETURNING,
             )
-        return
+        ):
+            try:
+                if mode == WorkMode.MODE_WORKING:
+                    trans_key = "pause_failed"
+                    await self.coordinator.async_send_command("pause_execute_task")
+                    await self.coordinator.async_request_iot_sync()
```
Comment (mikey0000) [inline custom_components/mammotion/lawn_mower.py:231]:
```suggestion
```

File: custom_components/mammotion/lawn_mower.py [hunk 3]
```diff
@@ -166,77 +166,103 @@ async def async_start_mowing(self, **kwargs: Any) -> None:
...
+                WorkMode.MODE_READY,
+                WorkMode.MODE_RETURNING,
             )
-        return
+        ):
+            try:
+                if mode == WorkMode.MODE_WORKING:
+                    trans_key = "pause_failed"
+                    await self.coordinator.async_send_command("pause_execute_task")
+
```
Comment (mikey0000) [inline custom_components/mammotion/lawn_mower.py:231]:
```suggestion
await self.coordinator.async_request_iot_sync()
```

**Pull request merged**